cirq-rigetti/cirq_rigetti/quil_input.py

```diff
@@ -234,50 +303,127 @@ def circuit_from_quil(quil: str) -> Circuit:
     References:
         https://github.com/rigetti/pyquil
     """
+    if isinstance(quil, str):
+        program = Program(quil)
+    else:
+        program = quil
     circuit = Circuit()
-    defined_gates = SUPPORTED_GATES.copy()
-    instructions = parse(quil)
 
-    for inst in instructions:
-        # Add DEFGATE-defined gates to defgates dict using MatrixGate.
-        if isinstance(inst, DefGate):
-            if inst.parameters:
-                raise UnsupportedQuilInstruction(
-                    "Parameterized DEFGATEs are currently unsupported."
-                )
-            defined_gates[inst.name] = MatrixGate(inst.matrix)
+    defined_gates, parameter_transformers = get_defined_gates(program)
+
+    kraus_model: Dict[Tuple[QubitDesignator, ...], List[NDArray[np.complex_]]] = {}
+    confusion_maps: Dict[int, NDArray[np.float_]] = {}
+
+    # Interpret the Pragmas
+    for inst in program:
+        if not isinstance(inst, Pragma):
+            continue
+
+        # ADD-KRAUS provides Kraus operators that replace the gate operation
+        if inst.command == "ADD-KRAUS":
+            args = inst.args
+            gate_name = str(args[0])
+            if gate_name in matrices.QUANTUM_GATES:
+                u = matrices.QUANTUM_GATES[gate_name]
+            elif gate_name in defined_gates:
+                u = defined_gates[gate_name]
+            else:
+                raise UndefinedQuilGate(f"{gate_name} is not known.")
+
+            entries = np.fromstring(
+                inst.freeform_string.strip("()").replace("i", "j"), dtype=np.complex_, sep=" "
+            )
+            dim = int(np.sqrt(len(entries)))
+            kraus_gate_op = entries.reshape((dim, dim))
 
+            kraus_op = remove_gate_from_kraus([kraus_gate_op], u)[0]
+
+            if args in kraus_model:
+                kraus_model[args].append(kraus_op)
+            else:
+                kraus_model[args] = [kraus_op]
+
+        # READOUT-POVM provides a confusion matrix
+        elif inst.command == "READOUT-POVM":
+            qubit = qubit_index(inst.args[0])
+            entries = np.fromstring(
+                inst.freeform_string.strip("()").replace("i", "j"), dtype=np.float_, sep=" "
+            )
+            confusion_matrix = entries.reshape((2, 2)).T
+
+            """
+            Incompatible types in assignment
+                expression has type
+                    "ndarray[Any, dtype[complexfloating[Any, Any]]]"
+                target has type
+                    "ndarray[Any, dtype[floating[Any]]]"
+            """
+            confusion_maps[qubit] = confusion_matrix
```

Comment:
This one is confusing. The target is explicitly typed:
```py
confusion_maps: Dict[int, NDArray[np.float_]]
```

and `confusion_matrix` seems to think it has the following type due to the previous line:
```py
confusion_matrix: ndarray[Any, dtype[float_]]
````

Reply:
I can't explain this..

Reply:
I've added a type: ignore comment directive to this line.